File: gloSuite9023/gloSuite/gloCommon/gloUserControlLibrary/Classes/Cls_TabIndexSettings.vb

Imports System.Data
Imports System.Collections.Generic
Imports System.Text
Imports System.Windows.Forms
Imports System.Collections
Imports System.Collections.Specialized
Imports System.Diagnostics



Public Class Cls_TabIndexSettings

#Region " Class For Comparing two controls in the selected tab scheme. "

    Private Class TabSchemeComparer
        Implements IComparer
        Private comparisonScheme As TabScheme

#Region "IComparer Members"

        Public Function Compare(x As Object, y As Object) As Integer Implements IComparer.Compare
            Dim control1 As Control = TryCast(x, Control)
            Dim control2 As Control = TryCast(y, Control)

            If control1 Is Nothing OrElse control2 Is Nothing Then
                Debug.Assert(False, "Attempting to compare a non-control")
                Return 0
            End If

            If comparisonScheme = TabScheme.None Then
                Return 0
            End If

            If comparisonScheme = TabScheme.AcrossFirst Then
                ' The primary direction to sort is the y direction (using the Top property).
                ' If two controls have the same y coordination, then we sort them by their x's.
                If control1.Top < control2.Top Then
                    Return -1
                ElseIf control1.Top > control2.Top Then
                    Return 1
                Else
                    Return (control1.Left.CompareTo(control2.Left))
                End If
            Else
                ' comparisonScheme = TabScheme.DownFirst
                ' The primary direction to sort is the x direction (using the Left property).
                ' If two controls have the same x coordination, then we sort them by their y's.
                If control1.Left < control2.Left Then
                    Return -1
                ElseIf control1.Left > control2.Left Then
                    Return 1
                Else
                    Return (control1.Top.CompareTo(control2.Top))
                End If
            End If
        End Function

#End Region


        Public Sub New(scheme As TabScheme)
            comparisonScheme = scheme
        End Sub
    End Class

#End Region

#Region " Variable and Enum "

    Private container As Control
    Private schemeOverrides As Hashtable
    Private curTabIndex As Integer = 0

    Public Enum TabScheme
        None
        AcrossFirst
        DownFirst
    End Enum

#End Region

#Region " Public and Private Methods "

    Public Sub New(container As Control)
        Me.container = container
        Me.curTabIndex = 0
        Me.schemeOverrides = New Hashtable()
    End Sub

    Private Sub New(container As Control, curTabIndex As Integer, schemeOverrides As Hashtable)
        Me.container = container
        Me.curTabIndex = curTabIndex
        Me.schemeOverrides = schemeOverrides
    End Sub

    Public Sub SetSchemeForControl(c As Control, scheme As TabScheme)
        schemeOverrides(c) = scheme
    End Sub

    Public Function SetTabOrder(scheme As TabScheme) As Integer
        Try
            Dim controlArraySorted As New ArrayList()
            controlArraySorted.AddRange(container.Controls)
            controlArraySorted.Sort(New TabSchemeComparer(scheme))

            For Each c As Control In controlArraySorted
                Debug.WriteLine("Tab Index Settings:  Changing tab index for " & c.Name)


                c.TabIndex = System.Math.Max(System.Threading.Interlocked.Increment(curTabIndex), curTabIndex - 1)
                If c.Controls.Count > 0 Then
                    ' Control has children -- recurse.
                    Dim childScheme As TabScheme = scheme
                    curTabIndex = (New Cls_TabIndexSettings(c, curTabIndex, schemeOverrides)).SetTabOrder(childScheme)
                End If
            Next

            Return curTabIndex
        Catch e As Exception
            Debug.Assert(False, "Exception in SetTabOrder:  " & e.Message)
            Return 0
        End Try
    End Function

#End Region

End Class

File: gloSuite9023/gloSuite/gloCommon/gloAuditTrail/Cls_TabIndexSettings.cs
using System;
using System.Data;
using System.Collections.Generic;
using System.Text;
using System.Windows.Forms;
using System.Collections;
using System.Collections.Specialized;
using System.Diagnostics;


namespace gloCommon
{
    public class Cls_TabIndexSettings
    {

        #region " Class For Comparing two controls in the selected tab scheme. "

        private class TabSchemeComparer : IComparer
        {
            private TabScheme comparisonScheme;

            #region IComparer Members

            public int Compare(object x, object y)
            {
                Control control1 = x as Control;
                Control control2 = y as Control;

                if (control1 == null || control2 == null)
                {
                    Debug.Assert(false, "Attempting to compare a non-control");
                    return 0;
                }

                if (comparisonScheme == TabScheme.None)
                {
                    return 0;
                }

                if (comparisonScheme == TabScheme.AcrossFirst)
                {
                    // The primary direction to sort is the y direction (using the Top property).
                    // If two controls have the same y coordination, then we sort them by their x's.
                    if (control1.Top < control2.Top)
                    {
                        return -1;
                    }
                    else if (control1.Top > control2.Top)
                    {
                        return 1;
                    }
                    else
                    {
                        return (control1.Left.CompareTo(control2.Left));
                    }
                }
                else    // comparisonScheme = TabScheme.DownFirst
                {
                    // The primary direction to sort is the x direction (using the Left property).
                    // If two controls have the same x coordination, then we sort them by their y's.
                    if (control1.Left < control2.Left)
                    {
                        return -1;
                    }
                    else if (control1.Left > control2.Left)
                    {
                        return 1;
                    }
                    else
                    {
                        return (control1.Top.CompareTo(control2.Top));
                    }
                }
            }

            #endregion


            public TabSchemeComparer(TabScheme scheme)
            {
                comparisonScheme = scheme;
            }
        }

        #endregion

        #region " Variable and Enum "

        private Control container;
        private Hashtable schemeOverrides;
        private int curTabIndex = 0;

        public enum TabScheme
        {
            None,
            AcrossFirst,
            DownFirst
        }

        #endregion

        #region " Public and Private Methods "

        public Cls_TabIndexSettings(Control container)
        {
            this.container = container;
            this.curTabIndex = 0;
            this.schemeOverrides = new Hashtable();
        }

        private Cls_TabIndexSettings(Control container, int curTabIndex, Hashtable schemeOverrides)
        {
            this.container = container;
            this.curTabIndex = curTabIndex;
            this.schemeOverrides = schemeOverrides;
        }

        public void SetSchemeForControl(Control c, TabScheme scheme)
        {
            schemeOverrides[c] = scheme;
        }

        public int SetTabOrder(TabScheme scheme)
        {
            ArrayList controlArraySorted = null;
            try
            {
                controlArraySorted = new ArrayList();
                controlArraySorted.AddRange(container.Controls);
                controlArraySorted.Sort(new TabSchemeComparer(scheme));

                foreach (Control c in controlArraySorted)
                {
                    Debug.WriteLine("Tab Index Settings:  Changing tab index for " + c.Name);


                    c.TabIndex = curTabIndex++;
                    if (c.Controls.Count > 0)
                    {
                        // Control has children -- recurse.
                        TabScheme childScheme = scheme;
                        curTabIndex = (new Cls_TabIndexSettings(c, curTabIndex, schemeOverrides)).SetTabOrder(childScheme);
                    }
                }

                return curTabIndex;
            }
            catch (Exception e)
            {
                Debug.Assert(false, "Exception in SetTabOrder:  " + e.Message);
                return 0;
            }
            finally
            {
                controlArraySorted = null;
            }
        }

        #endregion

    }
    
}
